### ISSUE TITLE
Filter items outputted by API

### ISSUE THREAD

**Issue by spee** [directus/app#2606]
I have a collection of items that have a `published_on` date, and I'm looking for a way to only output items whose publish date has passed. Think of this as a way to schedule posts, for example.

Currently, I give some items the status "Published" but with a date in the future, and I would like those items to stay effectively unpublished for the public (in the API's output). 

Am I right that the app doesn't offer a way to set this up? And/or can I (only) use a custom filter hook in the API for this?

I know this is the app's repo, but I am currently handling it with this hook:

```php
<?php

return [
  'filters' => [
    'item.read.posts' => function(\Directus\Hook\Payload $payload) {
      $posts = $payload->getData();

      foreach($posts as $index => $post) {
        $publish_date = new DateTime($post['published_on']);
        if($publish_date > new DateTime()) {
          unset($posts[$index]);
        }
      }

      $payload->replace($posts);

      return $payload;
    }
  ]
];
```

Items that get cleared this way result in a 404 response, which is nice. But it would still be a nice feature for the app to schedule publications.

**Comment by rijkvanzanten** [comment]
Any reason in particular you couldn't do this with the [`filter` query parameter](https://docs.directus.io/api/query/filter.html#filtering-by-dates-and-times)?

**Comment by spee** [comment]
Yes: the content must not be publicly available until the publish date.

I've also now realized that the filter hook prevents the app from getting access to pre-release items, so the aforementioned technique doesn't help me.

**Comment by rijkvanzanten** [comment]
Yeah, that's why I was wondering about that 🙂 The app uses the API directly, so forcing things to be hidden from the API will also hide them from the app.

The best thing you can do is try to extract if you're coming from the public role in that hook, and only return it

(@benhaynes another +1 for our new permission plans)

**Comment by spee** [comment]
I came back to this issue because the filter hook caused some other problems.

Instead, I've added an intermediate status for items so that they can be set to "ready to be published", and I use an action hook to perform a query to upgrade any "ready" items to "published" if their publication date is in the past. I've attached this action to the `application:boot` hook, because it didn't seem to be called on `item.read.[collection-name]:before`.
